File: src/main/java/net/nodus/global/config/DeletedDataCleanupService.java
package net.nodus.global.config;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import net.nodus.database.account.UserAccountRepository;
import net.nodus.database.auth.RefreshTokenRepository;
import net.nodus.database.site.SiteKeyRepository;
import net.nodus.database.site.SiteRepository;
import org.springframework.stereotype.Service;

@Service
public class DeletedDataCleanupService {

    private final SiteRepository siteRepository;
    private final SiteKeyRepository siteKeyRepository;
    private final UserAccountRepository userAccountRepository;
    private final RefreshTokenRepository refreshTokenRepository;

    public DeletedDataCleanupService(
        SiteRepository siteRepository,
        SiteKeyRepository siteKeyRepository,
        UserAccountRepository userAccountRepository,
        RefreshTokenRepository refreshTokenRepository
    ) {
        this.siteRepository = siteRepository;
        this.siteKeyRepository = siteKeyRepository;
        this.userAccountRepository = userAccountRepository;
        this.refreshTokenRepository = refreshTokenRepository;
    }

    public void hardDeleteExpiredData() {
        Instant threshold = Instant.now().minus(365, ChronoUnit.DAYS);

        siteRepository.deleteAll(siteRepository.findByDeletedAtBefore(threshold));
        siteKeyRepository.deleteAll(siteKeyRepository.findByDeletedAtBefore(threshold));
        userAccountRepository.deleteAll(userAccountRepository.findByDeletedAtBefore(threshold));
        refreshTokenRepository.deleteAll(refreshTokenRepository.findByDeletedAtBefore(threshold));
    }
}

File: src/main/kotlin/net/nodus/config/DeletedDataCleanupService.kt
package net.nodus.config

import net.nodus.account.UserAccountRepository
import net.nodus.auth.RefreshTokenRepository
import net.nodus.site.SiteKeyRepository
import net.nodus.site.SiteRepository
import org.springframework.stereotype.Service
import java.time.Instant
import java.time.temporal.ChronoUnit

@Service
class DeletedDataCleanupService(
    private val siteRepository: SiteRepository,
    private val siteKeyRepository: SiteKeyRepository,
    private val userAccountRepository: UserAccountRepository,
    private val refreshTokenRepository: RefreshTokenRepository,
) {
    fun hardDeleteExpiredData() {
        val threshold = Instant.now().minus(365, ChronoUnit.DAYS)

        siteRepository.deleteAll(siteRepository.findByDeletedAtBefore(threshold))
        siteKeyRepository.deleteAll(siteKeyRepository.findByDeletedAtBefore(threshold))
        userAccountRepository.deleteAll(userAccountRepository.findByDeletedAtBefore(threshold))
        refreshTokenRepository.deleteAll(refreshTokenRepository.findByDeletedAtBefore(threshold))
    }
}